src/main/resources/data/minecolonies/researches/civilian/1_firstaid.json

```diff
@@ -0,0 +1,27 @@
+{
+  "id": "firstaid",
+  "name": "First Aid",
+  "branch": "civilian",
+  "requiredUniversityLevel": 1,
+  "requiredItems":
```

Comment:
Code-side, it does default to minecraft as a namespace to match other behaviors in jsons, using the schema :

                    if (itemName.length == 2)
                    {
                        item = ForgeRegistries.ITEMS.getValue(new ResourceLocation(itemName[0], itemName[1]));
                    }
                    else if (itemName.length == 1)
                    {
                        item = ForgeRegistries.ITEMS.getValue(new ResourceLocation("minecraft", itemName[0]));
                    }

I've just included it for the existing items to be a clearer example for non-vanilla requirements, since very few current researches use non-vanilla items (mostly ancient tomes).

NBT was a possible use case (custom tools, god items like GregTech's dynamic stuff), most flexible item damage values (eg a chief sword with >50% durability), and also just if pack makers wanted to support multiple itemstacks of the same item instead of ItemStack(128) or ItemStack(75).  But the biggest reason was just parallelism: if requiredBuildings and effects are arrays of objects, it's a little clearer to end users if requiredItems does the same.  There's very little way to give feedback regarding malformed jsons since forge doesn't even report them up, so trying to make it as copy-pastable is ideal.

I could swap all of them, since both requiredBuildings and effects match the tag:property pattern, but input sanitization get messier.  Up to you all what you' think's gonna be saner, though.

Reply:
For the NBT stuff, it might make sense to use the functionality I added to parsing the ResourceLocation for crafterrecipes that understands an extended syntax with NBT support.

Reply:
![image](https://user-images.githubusercontent.com/39036909/100969958-fe24dc00-3501-11eb-8534-4f89f332c7c4.png)
![image](https://user-images.githubusercontent.com/39036909/100970030-1ac11400-3502-11eb-94ef-524a26f9e7a6.png)

Does this look better (outside of the few joke descriptions)?  I'm not entirely sure on how we want to split the Name/Description/EffectDescription setup -- right now, the University only displays the Description (derived from ID) and Effects, which works, but is a little unintuitive, and it might be useful to have longer descriptions in the WindowUI and shorter names for dialogue text.  And it might be useful to distinguish effects descriptions, especially for more complex behaviors (the earlier discord example being Knowledge Of The End).

Reply:
Okay. I thought about it a bit now. I don't think the description will ever be different than the actual effect (because that's what matters.

So you can scrap the desc, and leave only the name (I'd make that id.name="") and the effect (I'd make this dependent on the effect ID and NOT on the research id though).

Reply:
If I am understanding Mekle right, you could now change the effects to their own files, and instead of
```json
"effects": {
    "additionModifier": "Health*6"
}
```
You have something along the lines of
```json
"effects": {
    "health": 6
}
```
and that "health" will be used as the id for an effect JSON object in another file, or even just as the effect id. Each effect would have its own lang key, regardless of how many research cards it was associated with. 

However this proceeds, I would avoid putting two distinct values in a single string if they are not a set of similar delimited values.

In terms of UI, here was what I had in mind:
Research would be an icon and title, then underneath would be the required resources lined up on the right, and the building requirement represented with the building's hut block and number on the left. Then the description (if we still wnat one) and effects would be listed in the research title's tooltip. Click the title to start, click again to cancel.
... thiat might be hard to visualise; I'll draw a diagram if I get time.

And as a side note: it's usually better to paste the code blocks instead of screenshots.

Reply:
I agree with the above. It should just be the effect id with the value. I'd go as far as doing "healthaddition": 6. To make it more verbose.

Reply:
Something along the lines of : 

civilian/6_indefatigability.json
> {
  "comment": "name: Indefatigability * description : Healing Saturation Min +%s",
  "id": "indefatigability",
  "branch": "civilian",
  "requiredUniversityLevel": 6,
  "parentResearch": "fortitude",
  "requiredItems":
  [
    "minecraft:golden_apple": 64
  ],
  "effects":
  [
    "healingSaturationAddition": 4
  ]
}

effects/healingSaturationAddition.json:
>{
  "id": "healingSaturationAddition",
  "type": "additionMultiplier",
  "levels":
  [
     -0.5,
     -1,
     -1.5,
     -2,
     -5,
  ]
}

@ShadowProtocol , that workflow makes sense, and also makes for a much more resilient solution for managing research with long effect descriptions.  Merkle001 has also requested non-parent research dependency, so it may be worth considering how to implement that, but it's neither likely to come up in the short term nor too bad if it becomes part of the tooltip as well.

Reply:
Overall, I'm liking where this is going. The one thing that grates a bit still is the way requirements are handled. Here's my thinking: 
```
"requirements":
[
{"item" : "minecraft:golden_apple", "quantity": 64 },
{"building" : "cook", "level": 3 }
],```

Do we need something in the effect that specifically identifies the attribute it modifies, to make them more generic?

Reply:
That's doable, as long as it's no issue to see

>"requirements":
[
{"item" : "minecolonies:gate_iron", "quantity": 64 },
{"item" : "minecolonies:ancienttome", "quantity": 2 },
{"building" : "builder", "level": 3 },
{"research" : "letitgrow" },
{"research" : "plant_2" },
]

for a more complex use case.  I would like to keep requiredUniversityLevel separate from the requirements tags, simply because it has such different side effects.

> Do we need something in the effect that specifically identifies the attribute it modifies, to make them more generic?

I was intending some way to identify the attribute an effect modifies before the GlobalResearches are created (even if that means just indexing all effect files to get types), but strictly speaking it's not necessary.  The actual IResearchEffect extenders are mostly there to predict return on getEffect, and to aid localization between :

>   "com.minecolonies.coremod.research.effect.modifier.multiplication": "%s by %d percent",
  "com.minecolonies.coremod.research.effect.modifier.addition": "%s + %d",
  "com.minecolonies.coremod.research.effect.unlock": "Unlocks %s",

If you want to keep that space for autogeneration of text, that "type" tag in the effects.json will be enough.  If you're okay with this collapsing down to just BooleanResearchEffect and LeveledResearchEffect, the difference between :

>"effects":
[
"healingSaturationAddition": 5
"unlockbuildingsifter" : true
]

Is enough and easily parsed with GSON, and the only reason I'd been moving to put the prefixes into the tag name was for readability rather than a logic requirement.

Reply:
That's exactly what I was thinking for requirements, and I'm fine with university level and parent separate, as they are more structural than requirements in many ways. 

As for the effects, I dislike mixing data and directive in property names. "unlockbuilding" : "sifter" makes the property names predictable, rather than something requireing parsing. 

In general, that's what I want to see for the effects definition, and why I was asking for "modifiedAttribute" as part of it, as "healingSaturationAddition" feels like a case of something that needs to be parsed to understand, instead of a simple 'enum'/registry lookup.

Reply:
Understood.  I'm intending for effect->ID to be only used as a string and never parsed beyond that, instead using effectType to handle casting if you all decide to keep the four separate IGlobalResearch variants.  

Mostly a question of how much to have available for humans reading through the things, either in data packs, making localization files, so on.  It works fine for the ResearchConstants and ResearchInitializer, but tracking down and understanding "saturation" or "cap" in en_lang without the code context might benefit from more verbose matters.

But it does admittedly come with cost of upkeep/writing.

Reply:
Right. I'm sorry, I think you're on the right path with the broken out effects. the unlockbuildingsifter:true is what threw me, I didn't originally get that the true was just "presence", so I was looking at IDs as embedded meaning rather than as an ID. 

I see some cases where we might want to have levels for unlockability affects, so a 1 rather than a true might be for the best, unless there is a compelling reason to use a boolean.

Reply:
Since I have to do input sanitization anyway, having different object types means that I could distinguish different effectCategories from their tag-properties without doing the lookup.  But if unlocks have even a small chance of using numeric, probably better to just parse and map out the effects first, and that'll also let me log references in researches to non-existent effects better anyway.

Mostly just don't want to have the same property data defined in multiple places, since that's a quick path to consistency problems.  But if there's only effectname:level in the research json, that's less of a problem.

Will have to make clear that the level depends on the effect file, but that's documentation.

Can swap to a less verbose option for effectIds.  Raycoms used healthaddition as a moderately verbose example above, and that's not a bad model.